presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/RemoveRedundantCastToVarcharInJoinClause.java

```diff
@@ -0,0 +1,170 @@
+import static com.facebook.presto.sql.planner.PlannerUtils.addProjections;
+import static com.facebook.presto.sql.planner.plan.Patterns.join;
+import static com.facebook.presto.sql.relational.Expressions.castToBigInt;
+
+/**
+ * Remove redundant cast to varchar in join condition for queries like `select select * from orders o join customer c on cast(o.custkey as varchar) = cast(c.custkey as varchar)`
+ * Transform from
+ * <pre>
+ *     - Join
+ *          left_cast = right_cast
+ *          - Project
+ *              left_cast := cast(lkey as varchar)
+ *              - TableScan
+ *                  lkey BIGINT
+ *          - Project
+ *              right_cast := cast(rkey as varchar)
+ *              - TableScan
+ *                  rkey BIGINT
+ *
+ * </pre>
+ * into
+ * <pre>
+ *     - Join
+ *          new_lkey = new_rkey
+ *          - Project
+ *              left_cast := cast(lkey as varchar)
+ *              new_lkey := lkey
+ *              - TableScan
+ *                  lkey BIGINT
+ *          - Project
+ *              right_cast := cast(rkey as varchar)
+ *              new_rkey := rkey
+ *              - TableScan
+ *                  rkey BIGINT
+ * </pre>
+ * We will rely on optimizations later to remove unnecessary cast (if not used) and identity projection here.
+ * <p>
+ * Notice that we do not apply similar optimizations to queries with similar join condition like `cast(bigint as varchar) = varchar`. In general it can be converted to
+ * `bigint = try_cast(varchar as bigint)` as if the varchar here cannot be converted to bigint, try_cast will return null and will not match anyway. However, a special case is
+ * varchar begins with 0. `select cast(92 as varchar) = '092'` is false, but `select 92 = try_cast('092' as bigint)` returns true.
+ */
+public class RemoveRedundantCastToVarcharInJoinClause
+        implements Rule<JoinNode>
+{
+    private static final List<Type> TYPE_SUPPORTED = ImmutableList.of(INTEGER, BIGINT);
+    private final FunctionAndTypeManager functionAndTypeManager;
+    private final FunctionResolution functionResolution;
+
+    public RemoveRedundantCastToVarcharInJoinClause(FunctionAndTypeManager functionAndTypeManager)
+    {
+        this.functionAndTypeManager = functionAndTypeManager;
+        this.functionResolution = new FunctionResolution(functionAndTypeManager.getFunctionAndTypeResolver());
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isRemoveRedundantCastToVarcharInJoinEnabled(session);
+    }
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return join();
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        PlanNode leftInput = context.getLookup().resolve(node.getLeft());
+        PlanNode rightInput = context.getLookup().resolve(node.getRight());
+        if (!(leftInput instanceof ProjectNode) || !(rightInput instanceof ProjectNode)) {
```

Comment:
so that transformation won't work if the projection happens some levels down the tree?
for example
```
(select * from (select *, cast(key as varchar) as newkey) where ...) join T on newkey=cast(... as varchar)
```

Reply:
Correct, it will not work. I tried to do it, but it will complicated this rule as we not only need to extract this pattern traversing the tree, but also need to propagate the variables used in the cast expression (it not propagated already) to the join node. Since this rule already covered the query which motivated this optimization, I decided to do it in a simplified way for now.